<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>Primer-lint by Kiwwa</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/pygment_trac.css">
    <script src="javascripts/scale.fix.js"></script>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <h1 class="header">Primer-lint</h1>
        <p class="header">Primer analysis tool for the hiplex protocol</p>

        <ul>
          <li class="download"><a class="buttons" href="https://github.com/Kiwwa/primer-lint/zipball/master">Download ZIP</a></li>
          <li class="download"><a class="buttons" href="https://github.com/Kiwwa/primer-lint/tarball/master">Download TAR</a></li>
          <li><a class="buttons github" href="https://github.com/Kiwwa/primer-lint">View On GitHub</a></li>
        </ul>

        <p class="header">This project is maintained by <a class="header name" href="https://github.com/Kiwwa">Kiwwa</a></p>


      </header>
      <section>
        <hr><h2>
<a name="primer-lint-primer-analysis-tool-for-hiplex-primer-protocol" class="anchor" href="#primer-lint-primer-analysis-tool-for-hiplex-primer-protocol"><span class="octicon octicon-link"></span></a>Primer-Lint: Primer Analysis Tool for hiplex-primer protocol</h2>

<p>Version 1.0</p>

<p>Authors: Tu Nguyen-Dumont (1), Daniel J Park (1),
         Bernard J Pope (2), Luke D Shillabeer (2).</p>

<pre><code>     (1) Genetic Epidemiology Laboratory, Department of Pathology.
     (2) Victorian Life Sciences Computation Initiative.

     The University of Melbourne, Australia
</code></pre>

<p>Web:     <a href="https://github.com/kiwwa/primer-lint">https://github.com/kiwwa/primer-lint</a></p>

<p>License: BSD</p>

<p>Requirements: Python 2.7 and the BioPython Library.
(<a href="http://biopython.org">http://biopython.org</a>)</p>

<hr><h2>
<a name="general-description" class="anchor" href="#general-description"><span class="octicon octicon-link"></span></a>General description</h2>

<p>Primer-Lint was created to analyse the primers generated by the hiplex protocol 
(1) for a number of features that are believed to be either beneficial or 
detrimental to primer efficacy. These features can be used pre-PCR to detect
potentially troublesome primers or post-sequencing to analyse the impact each 
feature had on the primers efficacy. </p>

<hr><h2>
<a name="features-" class="anchor" href="#features-"><span class="octicon octicon-link"></span></a>Features </h2>

<p>Features able to be analysed for each primer are;</p>

<p>GC content: The count of G and C nucleotides within each given primer, reported 
            as a percentage.</p>

<ul>
<li>Example primer sequence is GGGAATAACC.</li>
<li>Total length of sequence is 10 nucleotides.</li>
<li>Count of G's and C's within sequence is 5.</li>
<li>GC content of example sequence is 5/10, 50%.</li>
</ul><p>Primer hairpins: Measures primer sequence self-interaction.</p>

<ul>
<li>Example sequence is TTGAATATAATGGCCCAATATTCC.</li>
<li>
<p>A run of six rungs of self-interaction are present within this sequence 
 (as shown below);</p>

<pre><code>                 TTGAATATAATG\
                   ||||||     G
                  CCTTATAACCC/
</code></pre>
</li>
</ul><p>Primer Dimer interactions: Measures two or more primers interaction with each 
other.</p>

<ul>
<li>Example sequences are:
  &gt; Sequence one - GGGATATACTAA
  &gt; Sequence two - AAAGCTCCCTAT</li>
<li>
<p>Sequences have six nuclotide run that could interact (as shown below);</p>

<pre><code>                    GGGATATACTAA
                    ||||||
              AAAGCTCCCTAT
</code></pre>
</li>
</ul><hr><h2>
<a name="command-line-interaction" class="anchor" href="#command-line-interaction"><span class="octicon octicon-link"></span></a>Command line interaction</h2>

<p>Interaction with the tool is via the command-line. </p>

<p>usage: primerlint.py [--importhiplex CSV] [--importfasta FAS] --outputlog O 
       [--cg] [--cgbrief] [--cgmin CGMIN] [--cgmax CGMAX] [--hairpin HP] 
       [--primerdimer PD]</p>

<p>Arguments: </p>

<p>--importhiplex
   Description: Imports a CSV file, formatted in the hiplex-primer (2) 
                output style. Expects primer-names in first column and sequence 
                in the second column.
   Type:        File</p>

<p>--importfasta
   Description: Imports a FASTA formatted file. Can process multiple FASTA 
                sequences from a single file.
   Type:        File</p>

<p>--outputlog
   Description: Location and name of the output file.
   Type:        File</p>

<p>--cg
   Description: Calculates the CG content percentage of each primer imported.
   Type:        Boolean</p>

<p>--cgbrief
   Description: Calculates the CG content percentage of each primer imported and
                generates a much briefer report.
   Type:        Boolean</p>

<p>--cgmin n
   Description: The minimum bound of acceptable CG content.
   Type:        Integer
   Default:     40</p>

<p>--cgmax n
   Description: The maximum bound of acceptable CG content.
   Default:     60
   Type:        Integer</p>

<p>--hairpin n
   Description: Determines all continuous hairpin locations on each primer 
                imported. Integer supplied to this argument is the minimum size 
                of hairpin to report.
   Default:     0
   Type:        Integer</p>

<p>--hairpinmaxsizes
   Description: Calculates the same hairpins as --hairpin, however generates a 
                significantly briefer report of only the maximum size of 
                hairpin, rather than all possible hairpins.
   Type:        Boolean</p>

<p>--primerdimer
   Description: Determines all possible primer-dimer interactions between all of
                the imported primers.
   Type:        Integer
   Default:     1
   Note:        As this is comparing all primers to all primers, the speed of 
                this calculation increases exponentially as more primers are 
                added. A maximum of 500 primers is recommended.</p>

<p>File type arguments can be specified as either relative or absolute paths. 
Relative paths are interpreted "relative" to the directory from which the 
program was executed, whereas absolute paths are not, and must be specified in 
full. If the file you wish to specify resides in the same directory from which 
the program is run then a relative path would be just the name of the file on 
its own. </p>

<p>Boolean type arguments do not require parameters; they will work simply by using
them.</p>

<p>Integer type arguments require an integer as a parameter. The use of the integer
for each argument has been specified within the arguments description above.</p>

<hr><h2>
<a name="inputs-and-outputs" class="anchor" href="#inputs-and-outputs"><span class="octicon octicon-link"></span></a>Inputs and Outputs</h2>

<p>Only one import file can be specified at a time, meaning both --importhiplex and
--importfasta cannot be used within a single run of the application. An error
message will be displayed if it is attempted.</p>

<p>Primer-Lint's inputs are:</p>

<ul>
<li>FASTA: The text after '&gt;' is used as a primer/region name, the sequence 
        underneath is used as the primer sequence.</li>
<li>CSV:   Column one is primer/region name, column two is the primer sequence.</li>
</ul><p>Both accepted input formats can contain multiple sequences within a single file.</p>

<p>Primer-Lint's output is:</p>

<ul>
<li><p>Log file: The log file will contain the generated information based on the 
           arguments provided when Primer-Lint is run. </p></li>
<li><p>An example of the output log is available at: <a href="http://pastebin.com/0xDvbpvH">http://pastebin.com/0xDvbpvH</a></p></li>
</ul><hr><h2>
<a name="example-usage-should-be-all-on-one-line" class="anchor" href="#example-usage-should-be-all-on-one-line"><span class="octicon octicon-link"></span></a>Example usage (should be all on one line)</h2>

<p>We want to analyse a set of 60 primers for their CG percentage, potential 
hairpins and primer-dimer interactions. </p>

<p>We believe CG percentage should be between 30 and 70, do not want to be notified
of any hairpin interaction that is 5 rungs or less or any primer-dimer 
interaction that is less than 5 rungs.</p>

<p>Here are the file locations;</p>

<p>File to import (relative location): /input/60-primers.csv
   File to export (relative location): /output/output-60-primers.log</p>

<p>This is the command that would fulfill those requirements;</p>

<p>python primer-lint.py --importhiplex input/60-primers.csv --outputlog 
   output/output-60-primers.log --cg --cgmin 30 --cgmax 70 --hairpin 6 
   --primerdimer 5</p>

<hr><h2>
<a name="references" class="anchor" href="#references"><span class="octicon octicon-link"></span></a>References</h2>

<ol>
<li><p><a href="http://www.biotechniques.com/BiotechniquesJournal/2013/August/A-high-plex-">http://www.biotechniques.com/BiotechniquesJournal/2013/August/A-high-plex-</a>
PCR-approach-for-massively-parallel-sequencing/biotechniques-345734.html</p></li>
<li><p><a href="https://github.com/bjpop/hiplex-primer">https://github.com/bjpop/hiplex-primer</a></p></li>
</ol>
      </section>
      <footer>
        <p><small>Hosted on <a href="http://pages.github.com">GitHub Pages</a> using the Dinky theme</small></p>
      </footer>
    </div>
    <!--[if !IE]><script>fixScale(document);</script><![endif]-->
		
  </body>
</html>
